src/expr/src/scalar/func/impls/timestamp.rs

```diff
@@ -61,38 +61,178 @@ sqlfunc!(
+        false
+    }
+
+    fn inverse(&self) -> Option<crate::UnaryFunc> {
+        None
+    }
+
+    fn is_monotone(&self) -> bool {
+        true
+    }
+}
+
+impl fmt::Display for CastTimestampToTimestamp {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("timestamp_to_timestamp")
+    }
+}
+
+#[derive(
+    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
+)]
+pub struct CastTimestampTzToTimestamp(pub Option<TimestampPrecision>);
+
+impl<'a> EagerUnaryFunc<'a> for CastTimestampTzToTimestamp {
+    type Input = CheckedTimestamp<DateTime<Utc>>;
+    type Output = Result<CheckedTimestamp<NaiveDateTime>, EvalError>;
+
+    fn call(
+        &self,
         a: CheckedTimestamp<DateTime<Utc>>,
     ) -> Result<CheckedTimestamp<NaiveDateTime>, EvalError> {
-        a.naive_utc().try_into().err_into()
+        let mut out = CheckedTimestamp::try_from(a.naive_utc())?;
+        out.round_to_precision(self.0);
+        Ok(out)
     }
-);
+
+    fn output_type(&self, input: ColumnType) -> ColumnType {
+        ScalarType::Timestamp { precision: self.0 }.nullable(input.nullable)
+    }
+
+    fn preserves_uniqueness(&self) -> bool {
+        //TODO (mouli) Refactor out separate types if precision is same, which will preserve uniqueness.
+        false
+    }
+
+    fn inverse(&self) -> Option<crate::UnaryFunc> {
+        to_unary!(super::CastTimestampToTimestampTz(self.0))
+    }
+
+    fn is_monotone(&self) -> bool {
+        true
+    }
+}
+
+impl fmt::Display for CastTimestampTzToTimestamp {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("timestamp_with_time_zone_to_timestamp")
+    }
+}
+
+#[derive(
+    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
+)]
+pub struct CastTimestampTzToTimestampTz(pub Option<TimestampPrecision>);
+
+impl<'a> EagerUnaryFunc<'a> for CastTimestampTzToTimestampTz {
+    type Input = CheckedTimestamp<DateTime<Utc>>;
+    type Output = Result<CheckedTimestamp<DateTime<Utc>>, EvalError>;
+
+    fn call(
+        &self,
+        mut a: CheckedTimestamp<DateTime<Utc>>,
+    ) -> Result<CheckedTimestamp<DateTime<Utc>>, EvalError> {
+        a.round_to_precision(self.0);
+        Ok(a)
+    }
+
+    fn output_type(&self, input: ColumnType) -> ColumnType {
+        ScalarType::TimestampTz { precision: self.0 }.nullable(input.nullable)
+    }
+
+    fn preserves_uniqueness(&self) -> bool {
+        //TODO (mouli) Refactor out separate types if precision is same, which will preserve uniqueness.
+        false
+    }
+
+    fn inverse(&self) -> Option<crate::UnaryFunc> {
+        None
+    }
+
+    fn is_monotone(&self) -> bool {
+        true
+    }
+}
+
+impl fmt::Display for CastTimestampTzToTimestampTz {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("timestamp_to_timestamp")
```

Comment:
timestamp_with_timezone_to_timestamp_with_timezone? Would seem consistent with the other names like `timestamp_to_timestamp_with_timezone`.

(Also uncovered)

Reply:
yes 🙈! thanks for catching